extension/persistence/relational-jdbc/src/main/java/org/apache/polaris/extension/persistence/relational/jdbc/DatasourceOperations.java

```diff
@@ -0,0 +1,184 @@
+              // re:throw this as unhandled exception
+              throw new RuntimeException(e);
+            }
+            sqlBuffer.setLength(0); // Clear the buffer for the next statement
+          }
+        }
+      }
+    } catch (IOException e) {
+      LOGGER.error("Error reading the script file", e);
+      throw new RuntimeException(e);
+    } catch (SQLException e) {
+      LOGGER.error("Error executing the script file", e);
+      throw e;
+    }
+  }
+
+  public <T, R> List<R> executeSelect(
+      @Nonnull String query,
+      @Nonnull Class<T> targetClass,
+      @Nonnull Function<T, R> transformer,
+      Predicate<R> entityFilter,
+      int limit)
+      throws SQLException {
+    try (Connection connection = borrowConnection();
+        Statement statement = connection.createStatement();
+        ResultSet s = statement.executeQuery(query)) {
+      return ResultSetToObjectConverter.collect(s, targetClass, transformer, entityFilter, limit);
+    } catch (SQLException e) {
+      LOGGER.error("Error executing query {}", query, e);
+      throw e;
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public int executeUpdate(String query) throws SQLException {
+    try (Connection connection = borrowConnection();
+        Statement statement = connection.createStatement()) {
+      return statement.executeUpdate(query);
+    } catch (SQLException e) {
+      LOGGER.error("Error executing query {}", query, e);
+      throw e;
+    }
+  }
+
+  public int executeUpdate(String query, Statement statement) throws SQLException {
+    LOGGER.debug("Executing query {} within transaction", query);
+    try {
+      return statement.executeUpdate(query);
+    } catch (SQLException e) {
+      LOGGER.error("Error executing query {}", query, e);
+      throw e;
+    }
+  }
+
+  public void runWithinTransaction(TransactionCallback callback) throws SQLException {
+    Connection connection = null;
+    try {
+      connection = borrowConnection();
+      connection.setAutoCommit(false); // Disable auto-commit to start a transaction
+
+      boolean result;
+      try (Statement statement = connection.createStatement()) {
+        result = callback.execute(statement);
+      }
+
+      if (result) {
+        connection.commit(); // Commit the transaction if successful
+      } else {
+        connection.rollback(); // Rollback the transaction if not successful
+      }
+
+    } catch (SQLException e) {
+      if (connection != null) {
+        try {
+          connection.rollback(); // Rollback on exception
+        } catch (SQLException ex) {
+          LOGGER.error("Error rolling back transaction", ex);
+          throw e;
+        }
+      }
+      LOGGER.error("Caught Error while executing transaction", e);
+      throw e;
+    } finally {
+      if (connection != null) {
+        try {
+          connection.setAutoCommit(true); // Restore auto-commit
```

Comment:
If we get an exception that is not a `SQLException`, this will cause the Tx to commit, right? Is that intended?

Reply:
no, this is really good catch as this is added as the NOTE specifically if I change AutoCommit, I would run into this issue. tried to fix it, please have a pass.